src/bscPlugin/codeActions/XmlScopeCodeActionsProcessor.ts

```diff
@@ -0,0 +1,58 @@
+import type { CodeAction, Range } from 'vscode-languageserver';
+import { isBrsFile } from '../../astUtils/reflection';
+import { DiagnosticCodeMap } from '../../DiagnosticMessages';
+import type { BscFile, BsDiagnostic } from '../../interfaces';
+import util from '../../util';
+import type { XmlScope } from '../../XmlScope';
+
+// eslint-disable-next-line @typescript-eslint/no-extraneous-class
+export class XmlScopeCodeActionProcessor {
+    public constructor(
+        public scope: XmlScope,
+        public file: BscFile,
+        public range: Range,
+        public diagnostics: BsDiagnostic[],
+        public codeActions: CodeAction[]
+    ) {
+
+    }
+
+    public process() {
+        for (const diagnostic of this.diagnostics) {
+            if (diagnostic.code === DiagnosticCodeMap.callToUnknownFunction) {
+                this.importXmlScript(diagnostic);
+            }
+        }
+    }
+
+    public importXmlScript(diagnostic: BsDiagnostic) {
+        //functionName is stored on this specific diagnostic
+        const lowerFunctionName = (diagnostic as any).functionName.toLowerCase();
+
+        //find every file with this function defined
+        for (const key in this.scope.program.files) {
+            const file = this.scope.program.files[key];
+            if (isBrsFile(file)) {
+                //TODO handle namespace-relative function calls
+                const stmt = file.parser.references.functionStatementLookup.get(lowerFunctionName);
+                const slashOpenToken = this.scope.xmlFile.parser.ast.component?.ast.SLASH_OPEN?.[0];
```

Comment:
I'm still opposed to peeking at the internal AST :)

We should have a generic "add script" codeaction generator (might start a codeaction utils), which would try to append a script after the last script tags, then try to add it after the interface, and then try to add it at the beginning of the component's body

Reply:
I did it this way so we don't occasionally insert two consecutive whitespaces. It's easier to reason about the whitespace after an attribute. Although I do agree that we should add this to the very end of the tag, so I adjusted this to find the FINAL attribute first, and then after the `<component` object second if there are no attributes

Reply:
Great idea. I'll work on that.

Reply:
I'll do this in a separate PR.